app/src/main/java/com/github/se/studybuddies/viewModels/SharedTimerViewModel.kt

```diff
@@ -1,203 +1,332 @@
+              }
+            }
+
+            override fun onCancelled(error: DatabaseError) {
+              Log.e("TimerViewModel", "Failed to read timer", error.toException())
+            }
+          })
+    } ?: error("Group UID is not set. Call setup() with valid Group UID.")
+  }
+
+  private fun setupTimer(timeRemaining: Long) {
+    countDownTimer?.cancel()
+    countDownTimer =
+        object : CountDownTimer(timeRemaining, 1000) {
+          override fun onTick(millisUntilFinished: Long) {
+            _timerValue.value = millisUntilFinished
           }
 
-          override fun onCancelled(error: DatabaseError) {
-            // Optionally handle database read/write operation cancellations
-            Log.e(
-                "SharedTimerViewModel",
-                "Listening to timer updates was cancelled: ${error.message}")
+          override fun onFinish() {
+            _timerValue.value = 0
+            _timerEnd.value = true
+            isRunning = false
           }
         }
-    timerRef.addValueEventListener(timerEventListener!!)
+
+    if (isRunning) {
+
+      countDownTimer?.start()
+    }
+  }
+
+  fun startTimer() {
+    val newEndTime = System.currentTimeMillis() + _timerValue.value
+    isRunning = true
+    setupTimer(_timerValue.value)
+    isRunning = true
+    groupUID?.let { uid ->
+      viewModelScope.launch { databaseConnection.updateGroupTimer(uid, newEndTime, true) }
+    }
+  }
+
+  fun pauseTimer() {
+    isRunning = false
+    countDownTimer?.cancel()
+    groupUID?.let { uid ->
+      viewModelScope.launch() {
+        databaseConnection.updateGroupTimer(
+            uid, System.currentTimeMillis() + _timerValue.value, false)
+      }
+    }
   }
+
+  fun resetTimer() {
+    isRunning = false
+    _timerValue.value = 0
+    _timerEnd.value = false
+    countDownTimer?.cancel()
+    groupUID?.let { uid ->
+      viewModelScope.launch() { databaseConnection.updateGroupTimer(uid, 0L, false) }
+    }
+  }
+
+  // Adding time functions
+  fun addHours(hours: Long) {
+    val additionalTime = hours * 3600 * 1000 // Convert hours to milliseconds
+    updateTimer(additionalTime)
+  }
+
+  fun addMinutes(minutes: Long) {
+    val additionalTime = minutes * 60 * 1000 // Convert minutes to milliseconds
+    updateTimer(additionalTime)
+  }
+
+  fun addSeconds(seconds: Long) {
+    val additionalTime = seconds * 1000 // Convert seconds to milliseconds
+    updateTimer(additionalTime)
+  }
+
+  private fun updateTimer(additionalTime: Long) {
+    val newTime = _timerValue.value + additionalTime
+    if (newTime >= 0) {
+      _timerValue.value = newTime
+      if (isRunning) {
+        setupTimer(newTime)
+      }
+      groupUID?.let { uid ->
+        val newEndTime = System.currentTimeMillis() + newTime
+        viewModelScope.launch { databaseConnection.updateGroupTimer(uid, newEndTime, isRunning) }
+      }
+    }
+  }
+}
+
   /*
+    private val databaseConnection = DatabaseConnection()
```

Comment:
You have a lot of functions that are commented at the end of your file. If you don't need them anymore you should also probably removed them in this PR